Add a filter on wp_insert_user function regarding $user_pass (ticket #49639)

mklute101 opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->
This PR adds filter on user typed password before hashing the password.
It also includes error check for empty password and a backslash. Unit test included for both.
`npm run test:php -- --group 49639`

Trac ticket: [49639](https://core.trac.wordpress.org/ticket/49639)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

mklute101 committed:
commiting track version

mklute101 committed:
Replace usage of `strpos()` with `str_contains()`.

WordPress core includes a polyfill for `str_contains()` on PHP < 8.0 as of WordPress 5.9.

mklute101 committed:
removed use of wp_unslash as WordPress hasn't done any escaping to this point.

mklute101 committed:
update @ since tag to 6.7.0

mklute101 committed:
test for empty pass and slash

TimothyBJacobs commented:
Thanks for the pr @mklute101! We need to fix the PHPCS issues. You should be able to run something like this:

```
./vendor/bin/phpcbf src/wp-includes/user.php
```

mklute101 committed:
remove check for slash, as it seems to be allowed and caused this test failure
1) WP_Test_REST_Users_Controller::test_user_roundtrip_as_editor
Failed asserting that 500 is identical to 200.

mklute101 committed:
ran PHPCS

mklute101 committed:
remove the test for backslash

TimothyBJacobs commented on src/wp-includes/user.php:
```diff
@@ -2591,9 +2605,21 @@ function wp_update_user( $userdata ) {
+			return new WP_Error( 'empty_pre_hash_password', __( 'Empty password.' ) );
+		}
+
+		if ( false !== strpos( $pre_hash_password, '\\' ) ) {
+			return new WP_Error( 'illegal_pre_hash_password', __( 'Passwords may not contain the character "\\".' ) );
+		}
```
I think we want the update and insert to match, so we should remove this block here as well.
